File: 프로그래머스/2/150369. 택배 배달과 수거하기/택배 배달과 수거하기.swift
import Foundation

struct House {
    let number: Int
    var quantity: Int
}

func solution(_ cap:Int, _ n:Int, _ deliveries:[Int], _ pickups:[Int]) -> Int64 {
    var result: Int64 = 0
    var stk1 = [House]()
    var stk2 = [House]()

    for (i, d) in deliveries.enumerated() where d > 0 {
        stk1.append(House(number: i + 1, quantity: d))
    }
    for (i, p) in pickups.enumerated() where p > 0 {
        stk2.append(House(number: i + 1, quantity: p))
    }
    
    while !stk1.isEmpty || !stk2.isEmpty {
        let dist = max(stk1.last?.number ?? 0 , stk2.last?.number ?? 0) * 2
        result += Int64(dist)

        var c = cap

        while c > 0 && !stk1.isEmpty {
            let idx = stk1.count - 1
            if stk1[idx].quantity > c {
                stk1[idx].quantity -= c
                c = 0
            }
            else {
                c -= stk1[idx].quantity
                stk1.removeLast()
            }
        }
        c = cap

        while c > 0 && !stk2.isEmpty {
            let idx = stk2.count - 1
            if stk2[idx].quantity > c {
                stk2[idx].quantity -= c
                c = 0
            }
            else {
                c -= stk2[idx].quantity
                stk2.removeLast()
            }
        }
    }

    return result
}

File: 프로그래머스/2/150369. 택배 배달과 수거하기/택배 배달과 수거하기.cpp
#include <bits/stdc++.h>

using namespace std;

long long solution(int cap, int n, vector<int> deliveries, vector<int> pickups) {
    long long answer = 0;
    int i = n-1, j = n-1;
    // n개의 집을 모두 탐색. (역순으로)
    while(i >= 0  || j >= 0) {
        // 배달해야할 집 탐색
        while(i >= 0 && deliveries[i] == 0) i--;
        // 수거해야할 집 탐색
        while(j >= 0 && pickups[j] == 0) j--;
        
        // 이동거리 계산
        int dist = (max(i, j)+1) * 2;
        answer += dist;
        
        // 배달 처리
        int cnt = cap;
        while(i >= 0 && cnt > 0) {
            // i번째 집에 cnt 다 배달하는 경우
            if(deliveries[i] >= cnt) {
                deliveries[i] -= cnt;
                cnt = 0;
            }
            else {
                cnt -= deliveries[i];
                deliveries[i] = 0;
                i--;
            }
        }
        
        cnt = cap;
        while(j >= 0 && cnt > 0) {
            // i번째 집에 cnt 다 배달하는 경우
            if(pickups[j] >= cnt) {
                pickups[j] -= cnt;
                cnt = 0;
            }
            else {
                cnt -= pickups[j];
                pickups[j] = 0;
                j--;
            }
        }
    }
    return answer;
}